### PR TITLE
refactor(ui): refactor country code

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

refactor country code logic, introduce country-code utils to handle all the logics

image

1. remove the country name display for now
2. extract and detect the country code from the i18n language suffix
3. if no valid country code found, search for the local hardcoded  i18n language mapped country code
4. if no valid found, fallback to 'US'
5. countryList should always place default detected countryCallingCode at the first place
6. countryList should filter the duplicate countryCallingCode
7. countryList should ordered by countryCallingCode

## Testing

test locally
UT case added

@logto-io/eng

### CODE REVIEW

File: packages/ui/src/utils/country-code.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+export type CountryMetaData = {
+  countryCode: CountryCode;
+  countryCallingCode: CountryCallingCode;
+};
+
+export const getCountryList = (): CountryMetaData[] => {
+  const defaultCountryCode = getDefaultCountryCode();
+  const defaultCountryCallingCode = getCountryCallingCode(defaultCountryCode);
+
+  // eslint-disable-next-line @silverhand/fp/no-mutating-methods
```
Comment (gao-sun) [inline packages/ui/src/utils/country-code.ts:54]: see the error message. always use `.slice` before `.sort`

File: packages/ui/src/utils/country-code.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  const countryList = getCountries()
+    .map((code) => ({
+      countryCode: code,
+      countryCallingCode: getCountryCallingCode(code),
+    }))
+    // Filter the detected default countryCode & duplicates
+    .filter(({ countryCallingCode }, index, self) => {
+      if (countryCallingCode === defaultCountryCallingCode) {
+        return false;
+      }
```
Comment (gao-sun) [inline packages/ui/src/utils/country-code.ts:63]: why return false directly here?

File: packages/ui/src/utils/country-code.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+export const getDefaultCountryCode = (): CountryCode => {
+  const { language } = i18next;
+  const detectedCountryCode = countryCallingCodeMap[language];
+
+  // Extract the country code from language tag suffix
+  const [, countryCode] = language.split('-');
+
+  if (countryCode && isValidCountryCode(countryCode)) {
+    return countryCode as CountryCode;
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts]: updated

File: packages/ui/src/utils/country-code.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  return detectedCountryCode ?? fallbackCountryCode;
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts]: updated

File: packages/ui/src/utils/country-code.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+} from 'libphonenumber-js';
+
+export const fallbackCountryCode = 'US';
+
+export const countryCallingCodeMap: Record<string, CountryCode> = {
+  zh: 'CN',
+  en: 'US',
+};
+
+export const isValidCountryCode = (countryCode: string) => {
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts]: updated

File: packages/ui/src/utils/country-code.ts (hunk 1 above)
Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts]: updated

File: packages/ui/src/utils/country-code.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  try {
+    getCountryCallingCode(countryCode as CountryCode);
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts:19]: added

File: packages/ui/src/utils/country-code.ts (hunk 2 above)
Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts:63]: should filter out if the calling code value is the same as the defaultCountryCallingCode.  

This is to make sure at the remove duplicate countryCallingCode step, we keep the exact default country entity and remove the rest.

Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts:63]: anyway this should be removed once we have the countryName back

File: packages/ui/src/utils/country-code.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        return false;
+      }
+
+      return (
+        self.findIndex((element) => element.countryCallingCode === countryCallingCode) === index
+      );
+    })
+    .slice()
+    // Sort by countryCallingCode
+    .sort((previous, next) => (next.countryCallingCode > previous.countryCallingCode ? -1 : 1));
```
Comment (demonzoo) [inline packages/ui/src/utils/country-code.ts:71]: nit: previous.countryCallingCode - next.countryCallingCode

**Review by demonzoo — APPROVED** [review]

File: packages/ui/src/utils/country-code.ts (hunk 2 above)
Comment (gao-sun) [reply packages/ui/src/utils/country-code.ts:63]: ok

**Review by gao-sun — APPROVED** [review]

File: packages/ui/src/utils/country-code.ts (hunk 7 above)
Comment (gao-sun) [reply packages/ui/src/utils/country-code.ts:71]: just realize it's number. +1 for this

Comment (simeng-li) [reply packages/ui/src/utils/country-code.ts:71]: string type not allowed

**Pull request merged**